adapters/ogury/ogury.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+func Builder(_ openrtb_ext.BidderName, config config.Adapter, _ config.Server) (adapters.Bidder, error) {
+	return &adapter{endpoint: config.Endpoint}, nil
+}
+
+func (a adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := buildHeaders(request)
+
+	request.Imp = filterValidImps(request)
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Invalid request. assetKey/adUnitId or request.site.publisher.id required",
+		}}
+	}
+
+	var errors []error
+	for i, imp := range request.Imp {
+		var impExt, impExtBidderHoist map[string]json.RawMessage
+		// extract ext
+		if err := jsonutil.Unmarshal(imp.Ext, &impExt); err != nil {
+			return nil, append(errors, &errortypes.BadInput{
+				Message: "Bidder extension not provided or can't be unmarshalled",
+			})
+		}
+		// find Ogury bidder params
+		if bidder, ok := impExt[openrtb_ext.PrebidExtBidderKey]; ok {
+			if err := jsonutil.Unmarshal(bidder, &impExtBidderHoist); err != nil {
+				return nil, append(errors, &errortypes.BadInput{
+					Message: "Ogury bidder extension not provided or can't be unmarshalled",
+				})
+			}
+		}
+
+		// extract every value from imp[].ext.bidder to imp[].ext
+		for key, value := range impExtBidderHoist {
+			impExt[key] = value
+		}
+
+		ext, err := jsonutil.Marshal(impExt)
+		if err != nil {
+			return nil, append(errors, &errortypes.BadInput{
+				Message: "Error while marshaling Imp.Ext bidder exension",
+			})
+		}
+		request.Imp[i].Ext = ext
+
+		// save adUnitCode
+		request.Imp[i].TagID = imp.ID
+	}
+
+	// currency conversion
+	for i, imp := range request.Imp {
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != "USD" {
+
+			// Convert to US dollars
+			convertedValue, err := requestInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "USD")
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			// Update after conversion. All imp elements inside request.Imp are shallow copies
+			// therefore, their non-pointer values are not shared memory and are safe to modify.
+			request.Imp[i].BidFloorCur = "USD"
+			request.Imp[i].BidFloor = convertedValue
+		}
+	}
+
+	requestJSON, err := jsonutil.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+
+}
+
+func filterValidImps(request *openrtb2.BidRequest) (validImps []openrtb2.Imp) {
```

Comment:
You may want to return an array of the results of unmarshling the imp.ext's since you unmarshal them again in the calling function. It is true that you unmarshal `imp.ext.bidder` into different structs on the two passes, but you could either save the `imp.ext` unmarshal, or just use `map[string]json.RawMessage` target for both purposes, and then unmarshal the AssetKey and AdUnitID keys into strings as one last step. Unmarshalling JSON tends to be a CPU heavy process.